app/src/main/java/org/mozilla/focus/browser/HomeTileRecyclerView.kt

```diff
@@ -0,0 +1,184 @@
+import org.mozilla.focus.home.HomeTilePlaceholderGenerator
+import org.mozilla.focus.home.HomeTileScreenshotStore
+import org.mozilla.focus.telemetry.TelemetryWrapper
+import org.mozilla.focus.utils.FormattedDomain
+
+/**
+ * Duration of animation to show custom tile. If the duration is too short, the tile will just
+ * pop-in. I speculate this happens because the amount of time it takes to downsample the bitmap
+ * is longer than the animation duration.
+ */
+private const val CUSTOM_TILE_TO_SHOW_MILLIS = 200L
+private val CUSTOM_TILE_ICON_INTERPOLATOR = DecelerateInterpolator()
+
+class HomeTileAdapter(
+        private val uiLifecycleCancelJob: Job,
+        private val tiles: MutableList<HomeTile>,
+        private val loadUrl: (String) -> Unit
+) : RecyclerView.Adapter<TileViewHolder>() {
+
+    override fun onBindViewHolder(holder: TileViewHolder, position: Int) = with (holder) {
+        val item = tiles[position]
+        when (item) {
+            is BundledHomeTile -> {
+                onBindBundledHomeTile(holder, item)
+                setLayoutMarginParams(iconView, R.dimen.bundled_home_tile_margin_value)
+            }
+            is CustomHomeTile -> {
+                onBindCustomHomeTile(uiLifecycleCancelJob, holder, item)
+                setLayoutMarginParams(iconView, R.dimen.custom_home_tile_margin_value)
+            }
+        }.forceExhaustive
+
+        itemView.setOnClickListener {
+            loadUrl(item.url)
+            TelemetryWrapper.homeTileClickEvent(item)
+        }
+
+        val tvWhiteColor = ContextCompat.getColor(holder.itemView.context, R.color.tv_white)
+        itemView.setOnFocusChangeListener { v, hasFocus ->
+            val backgroundResource: Int
+            val textColor: Int
+            if (hasFocus) {
+                backgroundResource = R.drawable.home_tile_title_focused_background
+                textColor = tvWhiteColor
+                menuButton.visibility = View.VISIBLE
+            } else {
+                backgroundResource = 0
+                textColor = Color.BLACK
+                menuButton.visibility = View.GONE
+            }
+            titleView.setBackgroundResource(backgroundResource)
+            titleView.setTextColor(textColor)
+        }
+    }
+
+    private fun setLayoutMarginParams(iconView: View, tileMarginValue: Int) {
+        val layoutMarginParams = iconView.layoutParams as ViewGroup.MarginLayoutParams
+        val marginValue = iconView.resources.getDimensionPixelSize(tileMarginValue)
+        layoutMarginParams.setMargins(marginValue, marginValue, marginValue, marginValue)
+        iconView.layoutParams = layoutMarginParams
+    }
+
+    fun getItemAtPosition(position: Int): HomeTile? {
+        if (position > -1 && position < itemCount) {
+            return tiles[position]
+        }
+        return null
+    }
+
+    fun addItemToAdapter(homeTile: HomeTile) {
+        tiles.add(homeTile)
+        notifyItemInserted(tiles.lastIndex)
```

Comment:
Hmmm this still has the problem of hard-coding the insertion. What if UX changes how the insertion works? I'd rather have this be correct for everything, rather than just fixed for now.

Thinking about this from a MVC perspective (and happy to do this IRL if you want), there are two things we want:
* No assumptions about *where* the tile is being inserted
* Efficient insertion (e.g., not invalidating the whole list)

After we pin the site, we can get the new cached list of tiles, right? My thought would be, can we do something here like `updateAdapterSingleInsertion` where the argument is the new list of tiles? That way we can make this more general, and the method can handle any "single insertion" adapter changes without the assumption of "new tiles get added to the end" being hidden in the method call.